### PR TITLE
Fix clearing subfigures

### LINKED ISSUE #22137: [Bug]: Cannot clear figure of subfigures
### Bug summary

Hi,
I would like to use `fig.clear()` on a figure that contains subfigures. This currently (as of Python 3.10.1, matplolib `3.6.0.dev1416+ga4f712079a.d20220107`) throws a ValueError.

### Code for reproduction

```python
import matplotlib.pyplot as plt
mainfig = plt.figure()
gs = mainfig.add_gridspec(ncols=1,nrows=1)
subfig = mainfig.add_subfigure(gs[0])
ax = subfig.subplots(1)
im = ax.imshow([[1, 2], [3, 4]])
mainfig.clear()
```

### Actual outcome

```
Traceback (most recent call last):
  File "/home/jay/ExtraDrive1/External/experiments/matplotlib/mwe/test_clear_figure_of_subfigures.py", line 7, in <module>
    mainfig.clear()
  File "/home/jay/ExtraDrive1/External/experiments/matplotlib/lib/matplotlib/figure.py", line 2782, in clear
    self.clf(keep_observers=keep_observers)
  File "/home/jay/ExtraDrive1/External/experiments/matplotlib/lib/matplotlib/figure.py", line 2760, in clf
    self.delaxes(ax)  # Remove ax from self._axstack.
  File "/home/jay/ExtraDrive1/External/experiments/matplotlib/lib/matplotlib/figure.py", line 930, in delaxes
    self._localaxes.remove(ax)
ValueError: list.remove(x): x not in list
```
### Expected outcome

The program should run and `mainfig.axes` as well as `mainfig.subfigs` should be empty.

### Additional information

I have fixed this in a local branch and am submitting a PR

### Operating system

Pop! OS 21.04

### Matplotlib Version

3.6.0.dev1416+ga4f712079a.d20220107

### Matplotlib Backend

QtAgg and module://matplotlib_inline.backend_inline

### Python version

Python 3.10.1

### Jupyter version

3.2.5

### Installation

git checkout

### PR DESCRIPTION
**Pull request by stanleyjs**
## PR Summary

Closes #22137 

``Figure.clf()`` and ``Figure.clear()`` were broken 
when ``Figure`` contained axes not stored in `Figure._localaxes`. #  Moved the
``clear`` and ``clf`` methods to ``FigureBase``, so ``Subfigure``s can now be
independently cleared. Added a routine to clear ``Figure.subfigs`` when 
``Figure.clear()`` is called. This effects the API as it removes active 
subfigure references when clearing a figure.     

This scenario occurred when ``Figure`` was instantiated with ``Subfigure``s.
The bug was thrown in a call to ``_localaxes.remove`` in ``FigureBase.delaxes``.
The children ``Subfigure``s do not register their axes in ``_localaxes``,
which meant that the ``_localaxes`` are empty during clear, leading to the bug.
This was fixed first by moving the ``.clear`` and ``.clf``
methods into ``FigureBase``, so that ``Subfigure`` would inherit these methods.
This allowed for ``Subfigure`` to clear its own axes, but because each child 
``Subfigure`` shares its ``_axstack`` with its parent ``Figure``, the ``clf``
function had to be modified: an extra call to ``_axstack.clear`` was removed.
This call had no effect on normal operation, as the stack is previously cleared
in ``clf`` by iterating over the ``_localaxes`` returned by ``FigureBase.axes``;
however, in the case of ``Subfigure.clear``, ``_axstack.clear`` cleared
the entire ``_axstack`` of the parent, thus wiping all axes independent of their
parent subfigure/figure. With these two changes (``clf/clear`` inheritance and 
removing the call to ``_axstack``), subfigures can now be cleared.  
However, the initial bug remained. To fix the bug in the clearing of the parent figure,
an additional loop was added to the beginning of ``clf``. This loop recurses through
child ``Subfigures``,clearing all children before returning to the original function.
As the child axes, which live in ``_axstack`` but not ``_localaxes``, are no longer in 
``Figure.axes``, the remainder of ``clf`` functions as before, clearing out
``Figure.axes`` without risk of calling ``delaxes`` on an axis that is not contained in
``_localaxes``. Finally, to make ``FigureBase.clf`` an exhaustive function that truely removes
all references to child objects, a line was added to clear ``Figure.subfigs`` of any
unlinked, cleared subfigures.

## PR Checklist

**Tests and Styling**
- [ x] Has pytest style unit tests (and `pytest` passes).
- [x ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ x] New features are documented, with examples if plot related.
- [ x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:684]: unrelated change

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:959]: subfigures cannot have toolbars, so this should be in the `Figure` class version of `clf`.

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:963]: is there any reason for this list to change?

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:2120]: please clean up the PEP8 errors

File: lib/matplotlib/tests/test_figure.py
Comment (jklymak) [inline lib/matplotlib/tests/test_figure.py:756]: Does clearing a sub fig remove it from the figure?  I'm not following why that is desirable?  The parent should remove it if we want it removed...

**Review by jklymak — COMMENTED** [review]
Mostly seems good.  However, I don't think clearing a subfigure should remove the subfigure (any more than clearing the top-level figure closes it).

File: doc/api/next_api_changes/behavior/22135-JSS3.rst
Comment (jklymak) [inline doc/api/next_api_changes/behavior/22135-JSS3.rst:33]: This note is too long!

File: lib/matplotlib/figure.py
Comment (stanleyjs) [reply lib/matplotlib/figure.py:963]: I wasn't sure if maybe a subfigure had a GUI widget tracking one of its axes maybe we needed to keep the subfigure around.

File: lib/matplotlib/tests/test_figure.py
Comment (stanleyjs) [reply lib/matplotlib/tests/test_figure.py:756]: Clearing  a subfigure should not remove it from the parent figure

File: doc/api/next_api_changes/behavior/22135-JSS3.rst
Comment (stanleyjs) [reply doc/api/next_api_changes/behavior/22135-JSS3.rst:33]: Yeah, sorry, I meant to leave this only in the PR.

File: lib/matplotlib/figure.py
Comment (stanleyjs) [reply lib/matplotlib/figure.py]: I removed this code, but unsure if this could introduce issues with callbacks. Not quite 100% on how they work.

File: lib/matplotlib/figure.py
Comment (stanleyjs) [reply lib/matplotlib/figure.py]: My IDE did not like the way that this was wrapping lines. Fixed

File: lib/matplotlib/figure.py
Comment (stanleyjs) [reply lib/matplotlib/figure.py]: Resolved

**Comment by stanleyjs** [comment]
Okay @jklymak thanks for your review, I have responded and fixed everything.  The only concern I have is the comment about whether or not the list of subfigures should be entirely cleared in the parent when the parent calls clear. As I mentioned, I am concerned about any callbacks when a GUI widget is tracking an axis in the subfigures.

**Comment by jklymak** [comment]
I would expect fig.clear to remove all children.  If there are callbacks that I expected to stay alive, I shouldn't clear the whole figure?

**Comment by stanleyjs** [comment]
Sounds fine to me @jklymak. The latest commit follows this convention.

**Comment by jklymak** [comment]
`/home/circleci/project/doc/api/next_api_changes/behavior/22135-JSS3.rst:2: WARNING: Title underline too short.`  ;-)

**Comment by stanleyjs** [comment]
Wow. Fixed.

@jklymak what tool checks for RST style?

**Comment by jklymak** [comment]
Our docs fail to build if there are warnings.

File: lib/matplotlib/tests/test_figure.py
Comment (tacaswell) [inline lib/matplotlib/tests/test_figure.py:691]: I have a mild preference to split each of these cases into its own test, but maybe that you can keep re-using the same figure is part of the test?

**Review by tacaswell — APPROVED** [review]
Left a minor comment about the test, but not willing to hold merging on that.

**Comment by tacaswell** [comment]
@stanleyjs Would you mind also squashing this to one commit (or we can do that on merge)?

**Comment by jklymak** [comment]
doc build is still unhappy...

File: doc/api/next_api_changes/behavior/22135-JSS3.rst
Comment (timhoffm) [inline doc/api/next_api_changes/behavior/22135-JSS3.rst:4]:
```suggestion
when ``Figure`` contained axes not stored in ``Figure._localaxes``.  Moved the
```

This should fix the doc build. The private attribute `_localaxes` does not have a rendered documentation and thus cannot be linked to, resulting in unhappy sphinx:

> /home/circleci/project/doc/api/next_api_changes/behavior/22135-JSS3.rst:3: WARNING: py:obj reference target not found: Figure._localaxes

**Comment by jklymak** [comment]
Hi @stanleyjs unfortunately this needs a rebase whereby you will need to resolve the conflicts in `figure.py`.  When you have done so and fixed the doc mistake, please ping me and I'll try an merge expeditiously.  If you need assistance with a rebase, feel free to ask here or on gitter.  Thanks!

**Comment by stanleyjs** [comment]
Sorry, did not mean to let this stagnate. I'll look into this ASAP

**Comment by stanleyjs** [comment]
@jklymak It did not appear to need an actual rebase, i was able to just resolve the merge conflicts. We'll see if it passes CI

**Comment by tacaswell** [comment]
@stanleyjs We prefer to rebase rather than merge the main branch into the feature branches.

**Comment by stanleyjs** [comment]
@tacaswell Very well. I definitely don't know how to rebase, but I guess there's a first time for everything

**Comment by stanleyjs** [comment]
@tacaswell @jklymak Can you point me in a direction on how to optimally use rebase? Or can you rebase it yourself? I just rebased it once, then the branch was 200 commits behind, so I merged it and tried to rebase the merge and the previous rebase into a single commit again and got stuck manually merging 200+ commits. I have no idea what to do.

**Comment by tacaswell** [comment]
Using rebasing definitly takes a few times through to have it make sense!  We have some documentation on rebasing at https://matplotlib.org/stable/devel/gitwash/development_workflow.html#rebasing-on-trunk and the best thing I have ever read about git is https://tom.preston-werner.com/2009/05/19/the-git-parable.html

When git suggests you merge the remote branch into yours when the push fails, it is giving you conservative but very bad advice!  At that point the correct path is `git push --force-with-lease remote branch_name`.

---

The steps I did here:

- went back to the commit before you merged master by clicking through the GH ui
- eventually I got to https://github.com/tacaswell/matplotlib/tree/6e4b1d9c45c413691a84edf04227cad039c78e29 which let me use the GH ui to create a branch on github at that commit (which I called "restore")
- I then checked out that branch locally 
- I rebased that branch onto origin/master
- I had to manually resolve conflicts on 3 commits (dace9f4624, 9dfcfd534b, 1d5ba460b1).  The first two were real conflicts that I think I resolved correctly, the third was whitespace.  At each broken step git left me with the merge confilct markers.  I use emacs + magit for this, but most IDEs have some sort of helper tool for dealing with this.
- I checked out  47c4a23ec4349c1742f3e399a0ad8db20469d389 and pushed that to my fork as `subfigure-clear` as a backup!
- I force-pushed my rebased branch to your branch (I can do this because of the "let maintainers push to this branch" box which it ticked in your UI)

Hopefully these notes are more helpful than confusing!

It might still be worth using `git rebase -i` to squash this all down to one commit any way.

**Comment by stanleyjs** [comment]
@tacaswell This seems to have stagnated again. I'd like to get it in main asap.

If I just resolve the changes in figure.py can it be integrated?

**Comment by tacaswell** [comment]
@stanleyjs can you try doing the rebase (you may want to use `rebase -i` to squash first) .

It looks like the test you added is failing

```
>       assert mainaxes in fig.axes
199
E       assert <AxesSubplot:> in []
200
E        +  where [] = <Figure size 640x480 with 0 Axes>.axes
201

202
lib/matplotlib/tests/test_figure.py:761: AssertionError

```

**Review by tacaswell — APPROVED** [review]

**Comment by tacaswell** [comment]
I'm not quite sure what happened here but somehow you had picked up changes on many many files.  The diff from the current branch to  cc71237 is:

```diff
diff --git a/lib/matplotlib/figure.py b/lib/matplotlib/figure.py
index 7ef783265d..57ea70597f 100644
--- a/lib/matplotlib/figure.py
+++ b/lib/matplotlib/figure.py
@@ -922,17 +922,16 @@ default: %(va)s
         self.suppressComposite = None
         self.callbacks = cbook.CallbackRegistry()
 
+        # first clear the axes in any subfigures
         for subfig in self.subfigs:
-            # first clear the axes in any subfigures
             subfig.clf(keep_observers=keep_observers)
+        self.subfigs = []
 
         for ax in tuple(self.axes):  # Iterate over the copy.
             ax.cla()
             self.delaxes(ax)  # Remove ax from self._axstack.
-
         self._axstack = _AxesStack()
         self.artists = []
-        self.subfigs = []
         self.lines = []
         self.patches = []
         self.texts = []
@@ -2166,6 +2165,7 @@ class Figure(FigureBase):
             naxes=len(self.axes),
         )
 
+    @_api.make_keyword_only("3.6", "facecolor")
     def __init__(self,
                  figsize=None,
                  dpi=None,
@@ -2356,7 +2356,7 @@ class Figure(FigureBase):
         # figure...
         for ax in self.axes:
             if hasattr(ax, '_colorbar'):
-                # colorbars list themselvs as a colorbar.
+                # colorbars list themselves as a colorbar.
                 return False
         return True
 
@@ -2850,10 +2850,6 @@ class Figure(FigureBase):
         if toolbar is not None:
             toolbar.update()
 
-    def clear(self, keep_observers=False):
-        """Clear the figure -- synonym for `clf`."""
-        self.clf(keep_observers=keep_observers)
-        
     @_finalize_rasterization
     @allow_rasterization
     def draw(self, renderer):

```

I meant to remove the subclass `clear` (the one from the parent class will do and correctly get the right `clf`) and I moved the resetting of `self.subfigs` up a bit to make it easier to read (which is subjective).

File: doc/api/next_api_changes/behavior/22135-JSS3.rst
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/22135-JSS3.rst:4]: `Figure._localaxes` is not public API, and so doesn't belong in the API change notes. I'm not sure that this change requires a note at all, tbh.

File: lib/matplotlib/figure.py
Comment (QuLogic) [inline lib/matplotlib/figure.py:921]: Please use [numpydoc](https://matplotlib.org/devdocs/devel/documenting_mpl.html#example-docstring) for docstrings.

File: lib/matplotlib/figure.py
Comment (QuLogic) [inline lib/matplotlib/figure.py:933]: I don't think this is correct, as `Figure` gets its own `_AxesStack`, but `SubFigure` uses its parent's, and replacing it here would break that link.

File: lib/matplotlib/tests/test_figure.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:745]: Why did you swap the order of the comparison here?

File: lib/matplotlib/tests/test_figure.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:751]: And here?

File: lib/matplotlib/tests/test_figure.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:743]: axes
```suggestion
    # e.1) removing just the axes leaves the subplot
```

File: lib/matplotlib/tests/test_figure.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:747]:
```suggestion
    # e.2) removing just the subaxes leaves the subplot
```

File: lib/matplotlib/tests/test_figure.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py:756]: There are only two; I think we can just write out both lines, since that corresponds with what happens after `clear`.
```suggestion
    assert mainaxes in fig.axes
    assert subaxes in fig.axes
```

**Comment by tacaswell** [comment]
I squashed down to one commit so we pass the PR cleanliness check (not add / remove a file).

File: lib/matplotlib/figure.py
Comment (QuLogic) [inline lib/matplotlib/figure.py:952]: Extra trailing quotes.

**Review by QuLogic — APPROVED** [review]

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:916]:
```suggestion
    def clear(self, keep_observers=False):
```

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:931]:
```suggestion
            subfig.clear(keep_observers=keep_observers)
```

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:953]:
```suggestion
```

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:2844]:
```suggestion
        super().clear(keep_observers=keep_observers)
```

File: lib/matplotlib/figure.py
Comment (jklymak) [inline lib/matplotlib/figure.py:2850]:
```suggestion
def clear(self, keep_observers=False):
          """Clear the figure -- synonym for `clf`."""
          self.clf(keep_observers=keep_observers)
```

**Review by jklymak — COMMENTED** [review]
One naming issue here.  `clf` means "clear last figure", but here we are using it to clear subfigures that were not necessarily the "last" in any meaningful sense.  I would argue that FigureBase should call this `FigureBase.clear` and call `subfig.clear()`, and that `Figure.clf` should alias to `Figure.clear`.  Otherwise, to me it is not clear if `Subfig.clf` should clear itself or clear the whole figure.

**Comment by tacaswell** [comment]
> One naming issue here. clf means "clear last figure", 

I have always read this as "CLear Figure" / "CLear Axes", but that aside, `Figures` have `clf` and if we want `SubFigure`s to be drop-ins for `Figure`s then `SubFigures` also need `clf`, but I'm on board with flipping the usage and alias direction to prefer `clear`.

**Comment by jklymak** [comment]
Oh, hmmm, well I looked up Matlab's docs and they just say "clear figure", so maybe you are right, and indeed `clf` takes a figure argument in Matlab, so I guess I am incorrect.

**Review by jklymak — APPROVED** [review]

**Comment by tacaswell** [comment]
I had a commit (not pushed) to fix the naming.  I'll make a new PR.

**Pull request merged**
